packages/runtime/src/next/client.ts

```diff
@@ -0,0 +1,135 @@
+  id: string
+  site: { id: string }
+  data: Element
+  snippets: {
+    id: string
+    code: string
+    location: 'HEAD' | 'BODY'
+    liveEnabled: boolean
+    builderEnabled: boolean
+    cleanup: string | null
+  }[]
+  fonts: { family: string; variants: string[] }[]
+  meta: {
+    title: string | null
+    description: string | null
+    keywords: string | null
+    socialImage: {
+      id: string
+      publicUrl: string
+      mimetype: string
+    } | null
+    favicon: {
+      id: string
+      publicUrl: string
+      mimetype: string
+    } | null
+  }
+  seo: {
+    canonicalUrl: string | null
+    isIndexingBlocked: boolean | null
+  }
+}
+
+export type MakeswiftPageSnapshot = {
+  document: MakeswiftPageDocument
+  origin: string
+  cacheData: NormalizedCacheObject
+}
+
+type MakeswiftConfig = {
+  origin?: string
+}
+
+export class Makeswift {
+  private apiKey: string
+  private origin: URL
+
+  constructor(apiKey: string, { origin = 'https://api.makeswift.com' }: MakeswiftConfig = {}) {
+    if (typeof apiKey !== 'string') {
+      throw new Error(
+        'The Makeswift client must be passed a valid Makeswift site API key: ' +
+          "`new Makeswift('<makeswift_site_api_key>')`\n" +
+          `Received "${apiKey}" instead.`,
+      )
+    }
+
+    this.apiKey = apiKey
+
+    try {
+      this.origin = new URL(origin)
+    } catch {
+      throw new Error(`The Makeswift client received an invalid origin parameter: "${origin}".`)
+    }
+  }
+
+  private async fetch(path: string, init?: RequestInit): Promise<Response> {
+    const response = await fetch(new URL(path, this.origin).toString(), {
+      ...init,
+      headers: { ...init?.headers, ['X-API-Key']: this.apiKey },
+    })
+
+    return response
+  }
+
+  async getPages(params: { path?: string } = {}): Promise<MakeswiftPage[]> {
+    const searchParams = new URLSearchParams(params)
+    const response = await this.fetch(`/v1/pages?${searchParams}`)
+
+    if (!response.ok) {
+      throw new Error(`Failed to get pages with error: "${response.statusText}"`)
+    }
+
+    const json = await response.json()
+
+    return json
+  }
+
+  private async createSnapshot(document: MakeswiftPageDocument): Promise<MakeswiftPageSnapshot> {
+    const client = new MakeswiftClient({ uri: new URL('graphql', this.origin).href })
+    const cacheData = await client.prefetch(document.data)
+
+    return { document, origin: this.origin.href, cacheData }
+  }
+
+  private async getPageSnapshotByPageId(
+    pageId: string,
+    { preview = false }: { preview?: boolean } = {},
+  ): Promise<MakeswiftPageSnapshot | null> {
+    const searchParams = new URLSearchParams({ preview: String(preview) })
+    const response = await this.fetch(`/v1/pages/${pageId}/document?${searchParams}`)
+
+    if (!response.ok) {
+      if (response.status === 404) return null
+
+      throw new Error(`Failed to get snapshot with error: "${response.statusText}"`)
+    }
```

Comment:
Can this receive a 401?

Thinking of a misconfigured API key - that might return a 404 though

Reply:
In that scenario it would throw an error with the message `Failed to get a snapshot with error: "Unauthorized"`, which is what I was going for. I only handle 404 so that the function returns `null` if the snapshot is not found

Reply:
Ok, cool. Just wanted to make sure you had considered that